### DISCUSSION TITLE
LCD thing again

### DISCUSSION THREAD

**Discussion by EAGrahamJr** [mattjlewis/diozero#119] [General]
Obviously too much time on my hands - https://github.com/EAGrahamJr/kobots/blob/main/adhoc/LCD%20Issues.md

I'll probably get around to figuring out what the differences are, but thought I'd at least leave this here as a note.

**Comment by mattjlewis** [comment]
Thank you for this, will incorporate but might take a few weeks. I've ordered both Freenove devices so that I can test locally.

**Comment by mattjlewis** [comment]
Looking into this - it might just be a trivial bug... Will get back to you ASAP.

**Reply by EAGrahamJr** [reply]
Take your time - fyi, I tweaked my "working version" of the LCD code in my repo to use the smaller delays between the enable-bit flip without issue. 

One of the consistent items was the `clear` command - everything indicated that it takes quite a bit longer to execute. If I leave out the 3ms delay, the LCD starts displaying incorrect characters - so that's obviously important. :grin:

**Comment by mattjlewis** [comment]
Made a really stupid error while refactoring this code 18 months ago - I wasn't saving the control bit info in the constructor. This is while final attributes are a good thing... Anyway, committed the fixes, need to think about releasing 1.3.5 soon.

**Reply by EAGrahamJr** [reply]
WORKS LIKE A CHAMP!!!!!!

**Comment by mattjlewis** [comment]
Really sorry to have caused so much both over this silly little error! I moved house recently and most of my Pi stuff was still in boxes. I wrote a test app to compare the I2C output between your SimpleLcd class and the one in diozero when I noticed that there was something odd. Then I spotted the initialisation error while assuming the error was related to different control pin configuration. Hey-ho, lesson learnt!

**Reply by EAGrahamJr** [reply]
Hey, not a problem! It was kinda fun getting into the bits and bytes again after a very long time.
